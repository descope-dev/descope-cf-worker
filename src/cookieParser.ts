export interface CookieOptions {
	expires?: string;
	maxAge?: number;
	domain?: string;
	path?: string;
	secure?: boolean;
	httpOnly?: boolean;
	sameSite?: "Strict" | "Lax" | "None";
	[key: string]: string | number | boolean | undefined;
}

interface Cookie {
	name: string;
	value: string;
	options?: CookieOptions;
}

/**
 * Parses a "Set-Cookie" header and returns an object containing the cookie name,
 * value, and options as per the HTTP standard.
 * https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie
 *
 * @param setCookieHeader - The "Set-Cookie" header string to parse.
 * @throws {Error} If the input does not match the expected format.
 * @returns An object with the cookie name, value, and options.
 */
export const deserialize = (setCookieHeader: string): Cookie => {
	const cookieRegex = /^([^=]+)=([^;]+);(?:\s*(.*))?$/;
	const match = setCookieHeader.match(cookieRegex);

	if (!match) {
		throw new Error("Invalid Set-Cookie header format");
	}

	const name = match[1];
	const value = match[2];
	const optionsString = match[3] || "";

	const options: CookieOptions = {};

	// Split options string by semicolons and trim whitespace
	const optionPairs = optionsString.split(";").map((option) => option.trim());

	for (const optionPair of optionPairs) {
		const [optionName, optionValue] = optionPair.split("=").map((part) => part.trim());

		if (optionName) {
			// Parse known options
			switch (optionName.toLowerCase()) {
				case "expires":
					options.expires = optionValue;
					break;
				case "max-age":
					options.maxAge = Number(optionValue);
					break;
				case "domain":
					options.domain = optionValue;
					break;
				case "path":
					options.path = optionValue;
					break;
				case "secure":
					options.secure = true;
					break;
				case "httponly":
					options.httpOnly = true;
					break;
				case "samesite":
					options.sameSite = optionValue as "Strict" | "Lax" | "None";
					break;
				default:
					// Unknown option, store as-is
					options[optionName] = optionValue;
			}
		}
	}

	return { name, value, options };
};

/**
 * Constructs a "Set-Cookie" header string from a cookie name, value, and options object
 * as per the HTTP standard.
 * https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie
 *
 * @param name - The name of the cookie.
 * @param value - The value of the cookie.
 * @param options - An object containing cookie options.
 * @returns A "Set-Cookie" header string.
 */
export const serialize = ({ name, value, options }: Cookie): string => {
	const parts: string[] = [];

	// Add the name and value to the header
	parts.push(`${name}=${value}`);

	if (options) {
		// Add known options to the header
		if (options.expires) {
			parts.push(`Expires=${options.expires}`);
		}
		if (options.maxAge !== undefined) {
			parts.push(`Max-Age=${options.maxAge}`);
		}
		if (options.domain) {
			parts.push(`Domain=${options.domain}`);
		}
		if (options.path) {
			parts.push(`Path=${options.path}`);
		}
		if (options.secure) {
			parts.push("Secure");
		}
		if (options.httpOnly) {
			parts.push("HttpOnly");
		}
		if (options.sameSite) {
			parts.push(`SameSite=${options.sameSite}`);
		}
	}

	// Add any other custom options to the header
	for (const key in options) {
		if (options[key] && !parts.includes(key) && options[key] !== undefined) {
			parts.push(`${key}=${options[key]}`);
		}
	}

	return parts.join("; ");
};
