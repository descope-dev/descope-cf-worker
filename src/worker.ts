import { deserialize, serialize } from "./cookieParser";

/* istanbul ignore next */
export default {
	// eslint-disable-next-line @typescript-eslint/no-unused-vars
	async fetch(request: Request, env: Env, ctx: ExecutionContext): Promise<Response> {
		return handleRequest(request, env);
	},
};

export const handleRequest = async (request: Request, env: Env) => {
	const url = new URL(request.url);
	const proxyUrl = new URL(url);
	proxyUrl.hostname = new URL(env.DESCOPE_BASE_URL).hostname;

	const res = await fetch(proxyUrl, request);
	const headers = new Headers(res.headers);
	const cookies = headers.getAll("set-cookie");
	headers.delete("set-cookie");
	cookies.forEach((value) => {
		const cookie = deserialize(value);
		switch (cookie.name) {
			case env.DESCOPE_SESSION_COOKIE:
			case env.DESCOPE_SESION_REFRESH_COOKIE:
				if (cookie.options) {
					cookie.options.domain = url.hostname;
					headers.append("x-descope-worker-proxy-cookie-modified", cookie.name);
				}
		}
		value = serialize(cookie);
		headers.append("set-cookie", value);
	});
	const cors = headers.get("access-control-allow-origin");
	if (cors && cors == new URL(env.DESCOPE_BASE_URL).origin) {
		headers.set("access-control-allow-origin", request.headers.get("origin") ?? cors);
	}
	headers.set("x-descope-worker-proxy-url", proxyUrl.href);
	return new Response(res.body, {
		...res,
		headers,
	});
};
