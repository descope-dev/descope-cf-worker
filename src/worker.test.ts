import { CookieOptions, serialize } from "./cookieParser";
import { handleRequest } from "./worker";

const env = {
	DESCOPE_BASE_URL: "https://api.descope.com",
	DESCOPE_SESSION_COOKIE: "DS",
	DESCOPE_SESION_REFRESH_COOKIE: "DSR",
};

describe("worker", () => {
	beforeEach(() => {
		const fetchMock = getMiniflareFetchMock();
		fetchMock.disableNetConnect();

		const DO_NOT_MODIFY = "DO_NOT_MODIFY";
		const value = "1234";
		const options: CookieOptions = {
			domain: new URL(env.DESCOPE_BASE_URL).hostname,
			expires: "1234",
			path: "/",
			secure: true,
			httpOnly: true,
			sameSite: "Strict",
		};

		fetchMock
			.get(env.DESCOPE_BASE_URL)
			.intercept({ path: "/" })
			.reply(200, "OK", {
				headers: {
					[DO_NOT_MODIFY]: DO_NOT_MODIFY,
					"Set-Cookie": [
						serialize({ name: env.DESCOPE_SESSION_COOKIE, value, options }),
						serialize({ name: env.DESCOPE_SESION_REFRESH_COOKIE, value, options }),
						serialize({ name: DO_NOT_MODIFY, value, options }),
					],
				},
			});
	});

	it("should alter descope cookie domain", async () => {
		const url = new URL(`https://auth.example.com/`);
		const response = await handleRequest(new Request(url), env);
		expect(response.headers).toMatchSnapshot();
	});

	it("should not alter response body", async () => {
		const url = new URL(`https://auth.example.com/`);
		const response = await handleRequest(new Request(url), env);
		expect(await response.text()).toBe("OK");
	});

	it("should work with all request methods", async () => {
		const fetchMock = getMiniflareFetchMock();

		const url = new URL(`https://auth.example.com/`);
		const methods = ["GET", "POST", "PUT", "DELETE", "PATCH", "OPTIONS", "HEAD"];
		methods.forEach((method) =>
			fetchMock
				.get(env.DESCOPE_BASE_URL)
				.intercept({ path: "/", method })
				.reply(200, "OK"),
		);
		const responses = await Promise.all(
			methods.map((method) => handleRequest(new Request(url, { method }), env)),
		);
		responses.forEach((response) => {
			expect(response.status).toBe(200);
		});
	});
});
