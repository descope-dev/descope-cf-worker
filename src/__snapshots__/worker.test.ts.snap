// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`worker should alter descope cookie domain 1`] = `
Headers {
  Symbol(headers list): HeadersList {
    "cookies": [
      "DS=1234; Expires=1234; Domain=auth.example.com; Path=/; Secure; HttpOnly; SameSite=Strict",
      "DSR=1234; Expires=1234; Domain=auth.example.com; Path=/; Secure; HttpOnly; SameSite=Strict",
      "DO_NOT_MODIFY=1234; Expires=1234; Domain=api.descope.com; Path=/; Secure; HttpOnly; SameSite=Strict",
    ],
    Symbol(headers map): Map {
      "do_not_modify" => {
        "name": "do_not_modify",
        "value": "DO_NOT_MODIFY",
      },
      "origin" => {
        "name": "origin",
        "value": "https://auth.example.com",
      },
      "set-cookie" => {
        "name": "set-cookie",
        "value": "DS=1234; Expires=1234; Domain=auth.example.com; Path=/; Secure; HttpOnly; SameSite=Strict, DSR=1234; Expires=1234; Domain=auth.example.com; Path=/; Secure; HttpOnly; SameSite=Strict, DO_NOT_MODIFY=1234; Expires=1234; Domain=api.descope.com; Path=/; Secure; HttpOnly; SameSite=Strict",
      },
      "x-descope-worker-proxy-cookie-modified" => {
        "name": "x-descope-worker-proxy-cookie-modified",
        "value": "DS, DSR",
      },
      "x-descope-worker-proxy-url" => {
        "name": "x-descope-worker-proxy-url",
        "value": "https://api.descope.com/v1/refresh",
      },
    },
    Symbol(headers map sorted): null,
  },
  Symbol(guard): "none",
  Symbol(realm): {
    "settingsObject": {},
  },
}
`;
